src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -22,25 +22,685 @@ internal enum Instruction
+    {
+        await System.Threading.Tasks.Task.CompletedTask;
+        yield return 3;
+        return 1;
+    }
+    async System.Collections.Generic.IAsyncEnumerable<int> M2()
+    {
+        return 4;
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_common });
+            comp.VerifyDiagnostics(
+                // (8,9): error CS1622: Cannot return a value from an iterator. Use the yield return statement to return a value, or yield break to end the iteration.
+                //         return 1;
+                Diagnostic(ErrorCode.ERR_ReturnInIterator, "return").WithLocation(8, 9),
+                // (10,60): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     async System.Collections.Generic.IAsyncEnumerable<int> M2()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M2").WithLocation(10, 60),
+                // (12,9): error CS1622: Cannot return a value from an iterator. Use the yield return statement to return a value, or yield break to end the iteration.
+                //         return 4;
+                Diagnostic(ErrorCode.ERR_ReturnInIterator, "return").WithLocation(12, 9)
+                );
+        }
+
+        [Fact]
+        public void AsyncIteratorWithReturnNull()
+        {
+            string source = @"
+class C
+{
+    async System.Collections.Generic.IAsyncEnumerable<int> M()
+    {
+        await System.Threading.Tasks.Task.CompletedTask;
+        yield return 3;
+        return null;
+    }
+    async System.Collections.Generic.IAsyncEnumerable<int> M2()
+    {
+        return null;
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_common });
+            comp.VerifyDiagnostics(
+                // (8,9): error CS1622: Cannot return a value from an iterator. Use the yield return statement to return a value, or yield break to end the iteration.
+                //         return null;
+                Diagnostic(ErrorCode.ERR_ReturnInIterator, "return").WithLocation(8, 9),
+                // (10,60): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     async System.Collections.Generic.IAsyncEnumerable<int> M2()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M2").WithLocation(10, 60),
+                // (12,9): error CS1622: Cannot return a value from an iterator. Use the yield return statement to return a value, or yield break to end the iteration.
+                //         return null;
+                Diagnostic(ErrorCode.ERR_ReturnInIterator, "return").WithLocation(12, 9)
```

Comment:
>ErrorCode.ERR_ReturnInIterator [](start = 27, length = 30)

Do we need to support the following?
```c#
async IAsyncEnumerable<int> M2()
{
    await Task.Delay(1);
    return M();
}
```

Reply:
I don't think so. You would write this instead:
```C#
async Task<IAsyncEnumerable<int>> M2()
{
  await Task.Delay(1);
  return M();
}
```